firmware-binaries/examples/hello/src/main.rs

```diff
@@ -7,29 +7,31 @@
 
 use ufmt::uwriteln;
 
-use bittide_sys::uart::Uart;
+mod wrappers;
 
 #[cfg(not(test))]
 use riscv_rt::entry;
 
-const STATUS_REG_ADDR: *mut u32 = 0xE000_0000 as *mut u32;
-const UART_ADDR: *mut u8 = 0xC000_0000 as *mut u8;
+use wrappers::*;
 
-fn test_success() {
-    unsafe {
-        STATUS_REG_ADDR.write_volatile(1);
+impl ufmt::uWrite for UART {
+    fn write_str(&mut self, s: &str) -> Result<(), Self::Error> {
+        for b in s.bytes() {
+            self.set_data(b);
+        }
+        Ok(())
     }
-}
 
-fn test_failure() {
-    unsafe {
-        STATUS_REG_ADDR.write_volatile(2);
-    }
+    type Error = ();
 }
 
 #[cfg_attr(not(test), entry)]
 fn main() -> ! {
-    let mut uart = unsafe { Uart::new(UART_ADDR) };
+    let DeviceInstances {
+        UART: mut uart,
+        StatusReg: status,
```

Comment:
Am I correct in understanding that `DeviceInstances` is used to auto generated components at compile time (e.g. `UART` based on the json description, and we automatically get generate the API (e.g. `data` based on the field and `set_data` to write?)

Reply:
Yes exactly, `DeviceInstances` contains all the different accessible components specified in the memory map! Every device type gets a Rust wrapper and every register/field of a device gets the appropriate getter and setter.